Implement PS-3879 (Encrypted temporary files)

percona-ysorokin opened the pull request:
https://jira.percona.com/browse/PS-3879

Introduced new global non-dynamic boolean variable 'encrypt_tmp_files' which,
when turned on, enables encryption of temporary files created for various
server needs (MyISAM 'OPTIMIZE' / 'REPAIR', filesort, etc.)
For each individual file a new random key is generated using 'my_rand_buffer()'
upon opening (no encryption / key storage plugin is involved in this process).
File data is encrypted using AES 256-bit algorithm in CBC block mode.

Cherry-picked commit MariaDB/server@d9340d6
"MDEV-8126 encryption for temp files"

Cherry-picked commit MariaDB/server@318c826
"always use my_b_pread() instead of mysql_file_pread()"

Cherry-picked commit MariaDB/server@6309a30
"my_b_fill, inline my_b_* functions instead of hairy macros"

Cherry-picked commit MariaDB/server@196e852
"misc IO_CACHE cleanups"

Cherry-picked commit MariaDB/server@1d21b22
"MDEV-10001 my_b_seek() may not work correctly after my_b_read() hits EOF"

Cherry-picked commit MariaDB/server@58b54b7
"MDEV-9044 : Binlog corruption in Galera"

Cherry-picked commit MariaDB/server@735a4a1
"MDEV-10508 Mariadb crash on out of disk space during dump import"

Cherry-picked commit MariaDB/server@1841557
"myisam/aria: don't mess with IO_CACHE::file"

Cherry-picked commit MariaDB/server@3a1c91d
"add P_S instrumentation to file operations in mf_iocache2.c"

Cherry-picked commit MariaDB/server@47d1679
"fix encryption.tempfiles to check that encrypt_tmp_files is ON"

Cherry-picked commit MariaDB/server@60dfe12
"MDEV-14868 MariaDB server crashes after using ROLLBACK TO when encrypt_tmp_files=ON"

Cherry-picked commit MariaDB/server@b75d845
"MDEV-14868 MariaDB server crashes after using ROLLBACK TO when encrypt_tmp_files=ON"

In contrast to MariaDB we preserve original 'my_b_fill()' implementation for
non-encrypted mode and use
'return _my_b_read(info, 0, 0) ? 0 : info->read_end - info->read_pos;'
only when 'MY_ENCRYPT' flag is set.

Applied experimental fix for MDEV-10259
"mysqld crash with certain statement length and order with Galera and encrypt-tmp-files=1"
(https://jira.mariadb.org/browse/MDEV-10259)
(http://lists.askmonty.org/pipermail/commits/2018-March/012144.html)
Fixed a problem in 'my_b_encr_read()' with updating 'read_end' and 'pos_in_file'.

Fixed problem in group_replication plugin code:
'Transaction_Message::append_cache()' method could call 'my_b_fill()' in case
when 'info->read_pos != info->read_end' which is not supported in the
encrypted mode.

Fixed "my_aes.h" interface to be plain C compatible.

Re-recorded 'main.mysqld--help-notwin' MTR test case because of the new
'--encrypt-tmp-files' system variable.

Fixed 'perfschema.show_sanity' MTR test case because of the new
'--encrypt-tmp-files' system variable.

Re-recorded 'perfschema.relaylog' and 'rpl.rpl_binlog_sender_fseek' MTR test
cases re-recorded because of the commit MariaDB/server@3a1c91d
"add P_S instrumentation to file operations in mf_iocache2.c".
Additional seek was already there, just not instrumented and
therefore, not reflected in statistics.

Added new 'sys_vars.enctypt_tmp_files_basic' MTR test case for the new
'encrypt_tmp_files' system variable.

Added 'mf_iocache-t' gunit test which tests basic 'io_cache' functionality
in encrypted / unencrypted mode.

Various formatting and spelling fixes.

percona-ysorokin commented:
Jenkins build link
http://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1035

laurynas-biveinis commented on mysql-test/t/percona_encrypt_tmp_files.test:
```diff
@@ -0,0 +1,104 @@
+
+if (`SELECT @@global.encrypt_tmp_files = 0`)
+{
+   --skip Test requires --encrypt-tmp-files
```
--die ?

percona-ysorokin replied:
Changed to `--die`

laurynas-biveinis commented on mysql-test/t/percona_encrypt_tmp_files.test:
```diff
@@ -0,0 +1,104 @@
+# Row binlog format to fill binlog cache faster
+--source include/have_binlog_format_row.inc
```
The test will be just skipped if the binlog format happens to be different (main suite does not use binlog format combinations) - maybe set the format in the testcase instead?

percona-ysorokin replied:
Good point

percona-ysorokin replied:
Reworked

laurynas-biveinis commented on mysql-test/t/percona_encrypt_tmp_files.test:
```diff
@@ -0,0 +1,104 @@
+# Row binlog format to fill binlog cache faster
+--source include/have_binlog_format_row.inc
+--source include/have_innodb.inc
+
```
--source include/have_myisam.inc

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -64,6 +65,18 @@ PSI_file_key key_file_io_cache;
+static int _my_b_cache_write_r(IO_CACHE *info, const uchar *Buffer,
+                               size_t Count);
+
+static io_cache_encr_read_function _my_b_encr_read= 0;
```
NULL

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -64,6 +65,18 @@ PSI_file_key key_file_io_cache;
+                               size_t Count);
+
+static io_cache_encr_read_function _my_b_encr_read= 0;
+static io_cache_encr_write_function _my_b_encr_write= 0;
```
NULL

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -97,6 +110,8 @@ static void
 init_functions(IO_CACHE* info)
 {
   enum cache_type type= info->type;
+  info->read_function= 0;                      /* Force a core if used */
```
NULL

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -97,6 +110,8 @@ static void
 {
   enum cache_type type= info->type;
+  info->read_function= 0;                      /* Force a core if used */
+  info->write_function= 0;                     /* Force a core if used */
```
NULL

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -106,21 +121,78 @@ init_functions(IO_CACHE* info)
+      info->write_function= _my_b_encr_write;
+      break;
+    }
+    /* fall through */
```
Why fall through here? Either read/write_function will be initialised already, either the DBUG_ASSERT below will fire

percona-ysorokin replied:
Reworked by separating "encrypted" / "non-encrypted" code paths to make it more readable.

laurynas-biveinis commented on sql/binlog.cc:
```diff
@@ -1025,7 +1025,8 @@ class Binlog_event_writer
         char checksum_buf[BINLOG_CHECKSUM_LEN];
         int4store(checksum_buf, checksum);
-        if (my_b_write(output_cache, checksum_buf, BINLOG_CHECKSUM_LEN))
+        if (my_b_write(output_cache, reinterpret_cast<uchar*>(checksum_buf),
```
static_cast not enough?

percona-ysorokin replied:
static_cast is not allowed for converting `char*` to `unsigned char*`

laurynas-biveinis commented on sql/filesort.cc:
```diff
@@ -1160,7 +1160,8 @@ write_keys(Sort_param *param, Filesort_info *fs_info, uint count,
   }
 
-  if (my_b_write(chunk_file, &merge_chunk, sizeof(merge_chunk)))
+  if (my_b_write(chunk_file, reinterpret_cast<uchar*>(&merge_chunk),
```
Likewise

percona-ysorokin replied:
The same: `Merge_chunk*` and `uchar*` are considered to be `unrelated types`, so reinterpret_cast only

laurynas-biveinis commented on sql/log_event.cc:
```diff
@@ -2170,7 +2170,7 @@ log_event_print_value(IO_CACHE *file, const uchar *ptr,
       my_timestamp_from_binary(&tm, ptr, meta);
       int buflen= my_timeval_to_str(&tm, buf, meta);
-      my_b_write(file, buf, buflen);
+      my_b_write(file, reinterpret_cast<uchar*>(buf), buflen);
```
Likewise

percona-ysorokin replied:
No cast is needed, `buf` is already `uchar*` here.

laurynas-biveinis replied:
buf is not uchar* ?

percona-ysorokin replied:
Fixed

laurynas-biveinis commented on sql/mf_iocache_encr.cc:
```diff
@@ -0,0 +1,251 @@
+  ulonglong counter;
+  uint block_length, last_block_length;
+  uchar key[MY_AES_BLOCK_SIZE];
+  ulonglong inbuf_counter;
```
Move after "counter" to save some alignment padding bytes?

percona-ysorokin replied:
Internal algorithm relies on the fact that `inbuf_counter` is the last member of the struct. Moreover, in this particular case there is no padding gaps:
```c++
ulonglong counter; // 8 bytes
uint block_length, last_block_length; // 4 + 4 = 8 bytes
uchar key[MY_AES_BLOCK_SIZE]; // 16 bytes
ulonglong inbuf_counter; // 8 bytes
```

robgolebiowski commented on sql/mf_iocache_encr.cc:
```diff
@@ -0,0 +1,251 @@
+
+    elength= wlength - (ebuffer - wbuffer);
+    set_iv(iv, pos_in_file, crypt_data->inbuf_counter);
+
```
IV should be encrypted first, before used. Based on this book https://books.google.pl/books?id=nnqdQsrZnJgC&lpg=PT141&ots=6GWXwmnDt8&dq=Bruce%20Schneier%20cbc%20or%20ctr&hl=pl&pg=PT133#v=onepage&q=Bruce%20Schneier%20cbc%20or%20ctr&f=false by Bruce Scheneier - point 4.3.4.3 The nonce should be encrypted before it should be used as an IV. Thus there need to be something like what Sergei included in his diff:
 my_aes_encrypt(iv, MY_AES_BLOCK_SIZE, eiv,
                 key, key_length, ecb_mode, NULL, false);
Where eiv - should be later used as IV.

percona-ysorokin commented:
Jenkins build links:
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1742/ (normal)
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1743/ (MTR args: --mysqld=--encrypt-tmp-files)

percona-ysorokin commented:
@laurynas-biveinis @robgolebiowski a new commit has been pushed to this PR and is ready for review.

laurynas-biveinis commented on include/my_sys.h:
```diff
@@ -487,46 +488,85 @@ typedef void (*my_error_reporter)(enum loglevel level, const char *format, ...)
 extern my_error_reporter my_charset_error_reporter;
 
-/* defines for mf_iocache */
+/* inline functions for mf_iocache */
```
Consider using MY_NODISCARD for all the changed functions (if it does not cause warnings in legacy code)

laurynas-biveinis commented on mysql-test/suite/binlog/t/binlog_check_flush_log_assertion.test:
```diff
@@ -50,7 +50,6 @@ SET GLOBAL debug='+d,error_unique_log_filename';
 RESET MASTER;
 
 SET GLOBAL debug='-d,error_unique_log_filename';
-SET @@GLOBAL.DEBUG= @debug_save;
```
Why?

percona-ysorokin replied:
Rolled back

laurynas-biveinis commented on mysql-test/t/percona_encrypt_tmp_files.test:
```diff
@@ -0,0 +1,126 @@
+INSERT t2 SELECT REPEAT(seq, 1000) FROM seq_31_to_40;
+COMMIT;
+
+# SHOW BINLOG EVENTS must sucseed without 'CORRUPTED BINLOG' error
```
succeed

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysql-test/t/percona_encrypt_tmp_files.test:
```diff
@@ -0,0 +1,126 @@
+SET GLOBAL binlog_cache_size = @saved_binlog_cache_size;
+
+let $MYSQLD_DATADIR = `SELECT @@datadir`;
+exec $MYSQL_BINLOG $MYSQLD_DATADIR/master-bin.000001 | $MYSQL;
```
I see that there is a RESET MASTER above, but make sure to test with --repeat=3 at least

percona-ysorokin replied:
```
./mysql-test/mtr --debug-server percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files percona_encrypt_tmp_files --parallel=8 --repeat=16
```
runs fine

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -72,7 +85,7 @@ PSI_file_key key_file_io_cache;
 
   NOTES
-    This is called on automaticly on init or reinit of IO_CACHE
+    This is called on automatically on init or reinit of IO_CACHE
```
Drop :(

percona-ysorokin replied:
???

laurynas-biveinis replied:
It's a comment typo fix. Only adds to diff maintenance

laurynas-biveinis replied:
I suppose you want to keep it then :)

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -1723,7 +1850,7 @@ int main(int argc, char** argv)
   if (total_bytes != status.st_size)
   {
-    fprintf(stderr,"Not the same number of bytes acutally  in file as bytes \
+    fprintf(stderr,"Not the same number of bytes actually  in file as bytes \
```
might as well remove the extra space between "actually" and "in"

percona-ysorokin replied:
fixed

laurynas-biveinis commented on mysys/mf_iocache2.c:
```diff
@@ -182,7 +183,6 @@ void my_b_seek(IO_CACHE *info,my_off_t pos)
   DBUG_VOID_RETURN;
 }
 
-
```
Drop

percona-ysorokin replied:
done

laurynas-biveinis commented on sql/binlog.cc:
```diff
@@ -5922,7 +5922,10 @@ bool MYSQL_BIN_LOG::reset_logs(THD* thd, bool delete_only)
     }
     // don't clear global_sid_map because it's used by the relay log too
     if (gtid_state->init() != 0)
+    {
```
Why? Is this something that should be reported upstream as a code quality bug at least?

percona-ysorokin replied:
Removed from this PR, will be reported later.

laurynas-biveinis commented on sql/binlog.cc:
```diff
@@ -5936,11 +5939,12 @@ bool MYSQL_BIN_LOG::reset_logs(THD* thd, bool delete_only)
+  if (error)
     name= const_cast<char*>(save_name);
+  else
+    my_free(const_cast<char *>(save_name));
```
Why? Is this something that should be reported upstream as a code quality bug at least?

percona-ysorokin replied:
Removed from this PR, will be reported later.

laurynas-biveinis commented on sql/mf_iocache_encr.cc:
```diff
@@ -0,0 +1,296 @@
+/*
+   Copyright (c) 2018, Percona Inc. All Rights Reserved.
```
Is this correct copyright notice for a MariaDB port? Or is this file not a MariaDB port?

percona-ysorokin replied:
Fixed

laurynas-biveinis commented on mysys/mf_iocache.c:
```diff
@@ -64,6 +65,18 @@ PSI_file_key key_file_io_cache;
 #define IO_ROUND_UP(X) (((X)+IO_SIZE-1) & ~(IO_SIZE-1))
 #define IO_ROUND_DN(X) ( (X)            & ~(IO_SIZE-1))
 
+static int _my_b_cache_read(IO_CACHE *info, uchar *Buffer, size_t Count);
```
MY_NODISCARD all of them

laurynas-biveinis commented on sql/mf_iocache_encr.cc:
```diff
@@ -0,0 +1,296 @@
+  memcpy(nonce_block, &position, sizeof(position));
+  memcpy(&nonce_block[sizeof(position)], &counter, sizeof(counter));
+
+  int encryption_result MY_ATTRIBUTE((unused))= my_aes_encrypt(nonce_block,
```
Consider checking in release build too

percona-ysorokin replied:
There is almost no change AES encrypt function can fail in ECB mode for a single block.

laurynas-biveinis commented on sql/sql_parse.cc:
```diff
@@ -4456,7 +4456,9 @@ case SQLCOM_PREPARE:
     } 
-    
+    else
+      res= 1; // reload_acl_and_cache failed
```
Why?

percona-ysorokin replied:
Reverted

laurynas-biveinis commented on mysql-test/suite/rpl/t/rpl_binlog_sender_fseek.test:
```diff
@@ -9,9 +9,9 @@
 # case will verify the P_S counter COUNT_MISC for "wait/io/file/sql/binlog"
 # EVENT_NAME at performance_schema.file_summary_by_event_name. In a single
-# binary log file case, the counter should not be more than 5 (as there are some
+# binary log file case, the counter should not be more than 6 (as there are some
```
Report the bug of PFS not registering seeks upstream

percona-ysorokin replied:
Reported PS-3937 "Some file operations in mf_iocache2.c are not instrumented"
https://jira.percona.com/browse/PS-3937

percona-ysorokin commented:
Rebased on the top of the latest 5.7 which includes IO_CACHE refactoring.

Jenkins build link (normal):
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1819/

Jenkins build link (--mysqld=--encrypt-tmp-files):
https://jenkins.percona.com/view/5.7/job/mysql-5.7-param/1826/

`sys_vars.encrypt_tmp_files_basic` in the second run fails constantly because of the default sys var value
changed with `--mysqld=--encrypt-tmp-files`.

percona-ysorokin commented:
@laurynas-biveinis The PR is ready for another round of review.

laurynas-biveinis commented on mysql-test/t/percona_encrypt_tmp_files.test:
```diff
@@ -0,0 +1,126 @@
+INSERT INTO t1(v) SELECT CONCAT(CHAR(ASCII('a') + s2.seq), REPEAT(' ', s1.seq))
+  FROM seq_0_to_9 AS s1, seq_0_to_26 AS s2;
+UPDATE t1 SET c = v, t = v;
+SELECT sql_big_result t, COUNT(t) FROM t1 GROUP BY t LIMIT 10;
```
sql_big_result uppercase
